src/agent/MessageSender.ts

```diff
@@ -21,18 +24,26 @@ class MessageSender {
   }
 
   public async sendMessage(outboundMessage: OutboundMessage): Promise<void> {
+    if (!this.outboundTransporter) {
+      throw new Error('Agent has no outbound transporter!')
+    }
+    const returnRoute = outboundMessage.payload.hasReturnRouting()
     const outboundPackage = await this.envelopeService.packMessage(outboundMessage)
-    await this.outboundTransporter.sendMessage(outboundPackage, false)
+    await this.outboundTransporter.sendMessage(outboundPackage, returnRoute)
   }
 
   public async sendAndReceiveMessage<T extends AgentMessage>(
     outboundMessage: OutboundMessage,
     ReceivedMessageClass: Constructor<T>
   ): Promise<InboundMessageContext<T>> {
+    if (!this.outboundTransporter) {
+      throw new Error('Agent has no outbound transporter!')
+    }
+
     outboundMessage.payload.setReturnRouting(ReturnRouteTypes.all)
 
     const outboundPackage = await this.envelopeService.packMessage(outboundMessage)
-    const inboundPackedMessage = await this.outboundTransporter.sendMessage(outboundPackage, true)
+    const inboundPackedMessage = await this.outboundTransporter.sendAndReceiveMessage(outboundPackage)
```

Comment:
I'm not fully understanding. Why did you add a `sendAndReceiveMessage` to the outbound transporter?

Reply:
Good question. I wanted to preserve the current behavior of `await agent.routing.downloadMessages()`. 

First, I thought that using `sendMessage(outbaoundPackage, true)` would be good enough. The problem is that in HTTP outbound we now pass the response immediately to the agent calling `this.agent.receiveMessage(wireMessage)`. It means that the `batch` response is not processed inside `downloadMessages` anymore. 

This is related to the problem we discussed in our last AFJ call about dispatching messages to the agent from a message handler. I listed as last point in the description:
"Update download messages functionality to process batch message and to remove sendAndReceiveMessage that is still implemented in the old way relaying on synchronicity of HTTP request-response."

I'm thinking if we could implement the solution as part of this PR. I usually tend to work in smaller batches. Maybe prioritize it and do it before WebSockets. Maybe it's not so big change 🤔